File: lua/claude-code/session.lua
-- A Session is one Claude conversation: its chat sidebar (which outlives the
-- process) and, while it's running, a sidecar process. A session that isn't
-- running — suspended while idle, exited, or opened from disk — is resumed
-- automatically when it's needed.

local Chat = require("claude-code.ui.chat")
local Permissions = require("claude-code.ui.permission")
local Sidecar = require("claude-code.sidecar")
local config = require("claude-code.config")
local control = require("claude-code.control")
local modes = require("claude-code.modes")
local tools = require("claude-code.ui.tools")

---@class claude_code.Session
---@field id string Claude session id (chosen up front for new sessions).
---@field title? string Custom title, or Claude Code's summary.
---@field cwd string
---@field chat claude_code.Chat
---@field busy boolean A turn is in progress.
---@field last_active integer os.time() of the last activity.
---@field mode? claude_code.PermissionMode Current permission mode (nil until Claude Code reports it).
---@field private started_mode? claude_code.PermissionMode Mode the process was started in.
---@field private persisted boolean Its transcript exists on disk (so it can be resumed).
---@field private sidecar? claude_code.Sidecar
---@field private permissions claude_code.Permissions
---@field private suspending boolean
---@field private pending_title? string Rename to apply once the transcript exists.
---@field private in_reply boolean The current turn already has a "Claude" header.
---@field private reply_has_text boolean
---@field private interrupted boolean
---@field private cost number Cumulative session cost reported by the last result.
local Session = {}
Session.__index = Session

local count = 0

--- Resuming shows at most this many of the most recent messages.
local REPLAY_LIMIT = 200

--- A random RFC 4122 v4 UUID.
local function uuid()
  math.randomseed(vim.uv.hrtime())
  return (
    ("xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx"):gsub("[xy]", function(c)
      return ("%x"):format(c == "x" and math.random(0, 15) or math.random(8, 11))
    end)
  )
end

---@class claude_code.SessionOpts
---@field title? string Name for a new session.
---@field info? table SDKSessionInfo of a stored session to resume.

---@param opts? claude_code.SessionOpts
---@return claude_code.Session?
function Session.new(opts)
  opts = opts or {}
  if not config.claude_path() then
    vim.notify("claude-code: `claude` executable not found; set `claude` in setup()", vim.log.levels.ERROR)
    return nil
  end

  count = count + 1
  local info = opts.info
  local self = setmetatable({
    id = info and info.sessionId or uuid(),
    title = info and (info.customTitle or info.summary) or opts.title,
    cwd = info and info.cwd or vim.fn.getcwd(),
    persisted = info ~= nil,
    busy = false,
    last_active = info and math.floor((info.lastModified or 0) / 1000) or os.time(),
    mode = config.options.permission_mode or modes.settings_default(info and info.cwd or vim.fn.getcwd()),
    suspending = false,
    in_reply = false,
    reply_has_text = false,
    interrupted = false,
    cost = 0,
  }, Session)
  self.chat = Chat.new({
    id = count,
    title = self.title,
    on_submit = function(text)
      return self:send(text)
    end,
    on_interrupt = function()
      self:interrupt()
    end,
    session_id = function()
      return self.id
    end,
    on_show = function()
      self.permissions:on_show()
    end,
    on_cycle_mode = function()
      self:set_mode(modes.next(self.mode))
    end,
  })
  self.permissions = Permissions.new(self.chat, function(id, answer)
    if self.sidecar then
      self.sidecar:send({
        type = "permission_response",
        id = id,
        behavior = answer.behavior,
        always = answer.always,
        updated_input = answer.updated_input,
        set_mode = answer.set_mode,
        message = answer.message,
      })
    end
  end, function()
    vim.notify(("Claude needs your input in “%s”"):format(self.title or "New session"), vim.log.levels.WARN)
  end)

  if info then
    self.chat:set_status({ activity = "Loading" })
    self:load_history()
  else
    self:start()
  end
  return self
end

function Session:running()
  return self.sidecar ~= nil and self.sidecar:running()
end

--- Waiting on the user (a permission card or question).
function Session:needs_attention()
  return self.permissions:pending()
end

--- Start (or restart) the process, resuming the transcript if there is one.
---@private
function Session:start()
  self.suspending = false
  local sidecar
  sidecar = Sidecar.new({
    on_event = function(event)
      -- Events can still arrive after the session was closed and its chat wiped.
      if self.sidecar == sidecar and self.chat:valid() then
        self:on_event(event)
      end
    end,
    on_exit = function(code, stderr)
      if self.sidecar ~= sidecar then
        return
      end
      self.sidecar = nil
      self.busy = false
      self.permissions:clear()
      self.chat:set_status({ activity = nil, stopped = self.suspending and "suspended" or "ended" })
      if code ~= 0 and not self.suspending then
        vim.notify(("claude-code: sidecar exited with code %d\n%s"):format(code, stderr), vim.log.levels.ERROR)
      end
    end,
  })
  self.sidecar = sidecar
  self.started_mode = self.mode
  self.chat:set_status({ activity = self.busy and "Thinking" or "Starting", stopped = false, mode = self.mode })
  sidecar:start({
    type = "init",
    cwd = self.cwd,
    claude_path = config.claude_path() --[[@as string]],
    model = config.options.model,
    -- Keeps the session's current mode across suspend/resume.
    permission_mode = self.mode,
    resume = self.persisted and self.id or nil,
    session_id = not self.persisted and self.id or nil,
    title = not self.persisted and self.title or nil,
  })
end

--- Make sure the process is running (e.g. after being suspended).
function Session:ensure_running()
  if not self:running() then
    self:start()
  end
end

--- Stop the process while idle; the conversation stays and resumes on demand.
function Session:suspend()
  if self:running() and not self.busy and not self:needs_attention() then
    self.suspending = true
    self.sidecar:stop()
  end
end

--- End the process (the transcript stays on disk and can be resumed).
function Session:stop()
  if self:running() then
    self.suspending = true
    if self.busy then
      -- Closing stdin alone would let the current turn run to completion first.
      self.sidecar:send({ type = "interrupt" })
    end
    self.sidecar:stop()
  end
end

--- Prompts that end the session, as in the Claude Code CLI.
local EXIT_COMMANDS = { ["/exit"] = true, ["/quit"] = true, ["exit"] = true }

---@param text string
---@return boolean sent
function Session:send(text)
  if EXIT_COMMANDS[vim.trim(text)] then
    require("claude-code.sessions").close(self)
    vim.notify(("claude-code: ended “%s”; resume it from :Claude sessions"):format(self.title or "New session"))
    -- Not "sent": the chat's buffers are gone, so there's no prompt to clear.
    return false
  end
  if self.busy then
    vim.notify("claude-code: Claude is still working; interrupt it first", vim.log.levels.WARN)
    return false
  end
  self:ensure_running()
  self.last_active = os.time()
  self.chat.transcript:user_message(text)
  self.busy = true
  self.in_reply = false
  self.reply_has_text = false
  self.interrupted = false
  self.chat:set_status({ activity = "Thinking" })
  self.sidecar:send({ type = "prompt", text = text })
  return true
end

function Session:interrupt()
  if self.busy and self:running() then
    self.interrupted = true
    self.chat:set_status({ activity = "Interrupting" })
    self.sidecar:send({ type = "interrupt" })
  end
end

--- Switch permission mode (takes effect immediately if running, else on start).
---@param mode claude_code.PermissionMode
function Session:set_mode(mode)
  if not modes.valid(mode) then
    vim.notify("claude-code: unknown permission mode: " .. tostring(mode), vim.log.levels.ERROR)
    return
  end
  if mode == "bypassPermissions" and self:running() and self.started_mode ~= "bypassPermissions" then
    -- The SDK only allows it for sessions started in it (a deliberate safety check).
    vim.notify(
      "claude-code: bypassPermissions has to be the starting mode; set `permission_mode` in setup()",
      vim.log.levels.ERROR
    )
    return
  end
  self.mode = mode
  self.chat:set_status({ activity = self.chat:activity(), mode = mode })
  if self:running() then
    self.sidecar:send({ type = "set_permission_mode", mode = mode })
  end
end

---@param title string
function Session:rename(title)
  self.title = title
  self.chat:set_title(title)
  if not self.persisted then
    -- No transcript to write the title to yet; apply it after the first turn.
    self.pending_title = title
    return
  end
  control.request("rename_session", { session_id = self.id, title = title, dir = self.cwd }, function(err)
    if err then
      vim.notify("claude-code: rename failed: " .. err, vim.log.levels.ERROR)
    end
  end)
end

---@private
function Session:reply()
  if not self.in_reply then
    self.chat.transcript:start_turn("assistant")
    self.in_reply = true
  end
end

--- Set the activity unless a permission card is waiting on the user.
---@private
---@param activity string
function Session:activity(activity)
  if not self.permissions:pending() then
    self.chat:set_status({ activity = activity })
  end
end

---@private
---@param event claude_code.SidecarEvent
function Session:on_event(event)
  self.last_active = os.time()
  if event.type == "ready" then
    if not self.busy then
      self.chat:set_status({ activity = nil })
    end
  elseif event.type == "sdk" then
    self:on_sdk_message(event.message --[[@as table]])
  elseif event.type == "permission_request" then
    self.permissions:request(event --[[@as claude_code.PermissionRequest]])
  elseif event.type == "permission_cancel" then
    self.permissions:cancel(event.id --[[@as integer]])
  elseif event.type == "error" then
    vim.notify("claude-code: " .. tostring(event.message), vim.log.levels.ERROR)
  end
end

---@private
---@param msg table An SDKMessage from @anthropic-ai/claude-agent-sdk.
function Session:on_sdk_message(msg)
  -- Subagent traffic is nested under a tool call; the top-level transcript skips it for now.
  if msg.parent_tool_use_id then
    return
  end
  local transcript = self.chat.transcript

  if msg.type == "system" and (msg.subtype == "init" or msg.subtype == "status") and msg.permissionMode then
    -- Claude Code reports the mode each turn, and when it changes (e.g. leaving plan mode).
    self.mode = msg.permissionMode
    self.chat:set_status({ activity = self.chat:activity(), mode = self.mode })
  end
  if msg.type == "system" and msg.subtype == "init" then
    self.chat:set_status({ activity = self.busy and "Thinking" or nil, model = msg.model })
  elseif msg.type == "stream_event" then
    local ev = msg.event
    if ev.type == "content_block_start" then
      local block = ev.content_block
      if block.type == "text" then
        self:reply()
        if self.reply_has_text then
          transcript:paragraph_break()
        end
        self:activity("Responding")
      elseif block.type == "thinking" then
        self:activity("Thinking")
      elseif block.type == "tool_use" then
        self:activity("Preparing " .. block.name)
      end
    elseif ev.type == "content_block_delta" and ev.delta.type == "text_delta" then
      self:reply()
      transcript:append(ev.delta.text)
      self.reply_has_text = true
    end
  elseif msg.type == "assistant" then
    -- Text already streamed in via stream_event; the full message is where tool calls are complete.
    for _, block in ipairs(msg.message.content or {}) do
      if block.type == "tool_use" then
        self:reply()
        transcript:tool_use(block.id, block.name, block.input or {})
        self:activity("Running " .. block.name)
      end
    end
  elseif msg.type == "user" then
    self:tool_results(msg.message and msg.message.content)
  elseif msg.type == "result" then
    self:finish_turn(msg)
  end
end

---@private
---@param content any user message content
function Session:tool_results(content)
  for _, block in ipairs(type(content) == "table" and content or {}) do
    if block.type == "tool_result" then
      self.chat.transcript:tool_result(
        block.tool_use_id,
        block.is_error == true and "error" or "success",
        tools.result_text(block.content)
      )
      self:activity("Thinking")
    end
  end
end

---@private
---@param msg table SDKResultMessage
function Session:finish_turn(msg)
  local transcript = self.chat.transcript
  transcript:cancel_pending_tools()

  local turn_cost = math.max((msg.total_cost_usd or self.cost) - self.cost, 0)
  self.cost = msg.total_cost_usd or self.cost
  local parts = {}
  local hl
  if self.interrupted then
    table.insert(parts, "Interrupted")
  elseif msg.subtype ~= "success" or msg.is_error then
    table.insert(parts, "Error: " .. (msg.subtype == "success" and tostring(msg.result) or msg.subtype))
    hl = "ClaudeCodeError"
  end
  table.insert(parts, ("%.1fs"):format((msg.duration_ms or 0) / 1000))
  if turn_cost > 0 then
    table.insert(parts, ("$%.4f"):format(turn_cost))
  end
  transcript:footer(table.concat(parts, " · "), hl)

  self.busy = false
  self.interrupted = false
  self.chat:set_status({ activity = nil, cost = self.cost })

  if not self.persisted then
    -- The transcript exists now: apply a rename made before the first turn, and
    -- pick up Claude Code's title if the session wasn't named.
    self.persisted = true
    if self.pending_title then
      local title = self.pending_title
      self.pending_title = nil
      self:rename(title)
    elseif not self.title then
      control.request("get_session_info", { session_id = self.id, dir = self.cwd }, function(_, session)
        if session and not self.title then
          self.title = session.customTitle or session.summary
          self.chat:set_title(self.title)
        end
      end)
    end
  end
end

-- Resuming ---------------------------------------------------------------------

--- User text that Claude Code records but doesn't show as something you typed.
---@param text string
local function hidden_user_text(text)
  return text:match("^%s*<[%w_-]+>") ~= nil -- <command-name>, <local-command-stdout>, <system-reminder>, ...
    or text:match("^Caveat: The messages below") ~= nil
end

--- Load the stored conversation into the transcript. The process starts when
--- the session is shown (or on send).
---@private
function Session:load_history()
  control.request("get_messages", { session_id = self.id, dir = self.cwd, tail = REPLAY_LIMIT }, function(err, result)
    if err or not result then
      vim.notify("claude-code: couldn't load session: " .. tostring(err), vim.log.levels.ERROR)
      self.chat:set_status({ activity = nil, stopped = "suspended" })
      return
    end
    self.chat.transcript:batch(function()
      self:replay(result.messages or {}, result.total or 0)
    end)
    if self:running() then
      self.chat:set_status({ activity = nil })
    else
      self.chat:set_status({ activity = nil, stopped = "suspended" })
    end
  end)
end

---@private
---@param messages table[] SessionMessage[]
---@param total integer
function Session:replay(messages, total)
  local transcript = self.chat.transcript
  if total > #messages then
    transcript:note(("%d earlier messages not shown"):format(total - #messages))
  end
  for _, m in ipairs(messages) do
    local content = type(m.message) == "table" and m.message.content
    if m.parent_tool_use_id then
      -- Subagent traffic: skipped, as in live sessions.
    elseif m.type == "user" then
      local texts = {}
      if type(content) == "string" then
        texts = { content }
      elseif type(content) == "table" then
        for _, block in ipairs(content) do
          if block.type == "text" then
            table.insert(texts, block.text)
          elseif block.type == "image" then
            table.insert(texts, "[image]")
          end
        end
      end
      local text = vim.trim(table.concat(texts, "\n"))
      if text:match("^%[Request interrupted by user") then
        transcript:footer("Interrupted")
      elseif text ~= "" and not hidden_user_text(text) then
        transcript:user_message(text)
        self.in_reply, self.reply_has_text = false, false
      end
      self:tool_results(content)
    elseif m.type == "assistant" and type(content) == "table" then
      for _, block in ipairs(content) do
        if block.type == "text" and block.text ~= "" then
          self:reply()
          if self.reply_has_text then
            transcript:paragraph_break()
          end
          transcript:append(block.text)
          self.reply_has_text = true
        elseif block.type == "tool_use" then
          self:reply()
          transcript:tool_use(block.id, block.name, block.input or {})
        end
      end
    end
  end
  transcript:cancel_pending_tools()
  self.in_reply, self.reply_has_text = false, false
  transcript:note(("↻ Resumed · last active %s"):format(os.date("%b %d %H:%M", self.last_active)))
end

return Session

File: sidecar/src/session.ts
// Session mode: one process per Claude conversation.

import { randomUUID } from "node:crypto";
import {
  query,
  type CanUseTool,
  type PermissionResult,
  type PermissionUpdate,
  type Query,
  type SDKUserMessage,
} from "@anthropic-ai/claude-agent-sdk";
import { onLines, write } from "./io.js";
import type { ImageAttachment, Inbound, InitRequest, Outbound } from "./protocol.js";

function send(event: Outbound): void {
  write(event);
}

/** An async queue of user messages feeding the SDK's streaming input mode. */
class Inbox implements AsyncIterable<SDKUserMessage> {
  private items: SDKUserMessage[] = [];
  private waiter: ((result: IteratorResult<SDKUserMessage>) => void) | undefined;
  private closed = false;

  push(text: string, shouldQuery = true, images: ImageAttachment[] = []): void {
    // Images go before the text, as the Messages API recommends.
    const content: SDKUserMessage["message"]["content"] =
      images.length === 0
        ? text
        : [
            ...images.map((image) => ({
              type: "image" as const,
              source: { type: "base64" as const, media_type: image.media_type, data: image.data },
            })),
            { type: "text" as const, text },
          ];
    const item: SDKUserMessage = {
      type: "user",
      message: { role: "user", content },
      parent_tool_use_id: null,
      ...(shouldQuery ? {} : { shouldQuery: false }),
    };
    if (this.waiter) {
      const resolve = this.waiter;
      this.waiter = undefined;
      resolve({ value: item, done: false });
    } else {
      this.items.push(item);
    }
  }

  close(): void {
    this.closed = true;
    this.waiter?.({ value: undefined, done: true });
    this.waiter = undefined;
  }

  [Symbol.asyncIterator](): AsyncIterator<SDKUserMessage> {
    return {
      next: () => {
        const item = this.items.shift();
        if (item) return Promise.resolve({ value: item, done: false });
        if (this.closed) return Promise.resolve({ value: undefined, done: true });
        return new Promise((resolve) => (this.waiter = resolve));
      },
    };
  }
}

const inbox = new Inbox();
interface PendingPermission {
  resolve: (result: PermissionResult) => void;
  suggestions?: PermissionUpdate[];
}

const pendingPermissions = new Map<number, PendingPermission>();
let nextPermissionId = 1;
let session: Query | undefined;
/** Neovim closed stdin; the session is shutting down on purpose. */
let closing = false;

const canUseTool: CanUseTool = (toolName, input, options) =>
  new Promise((resolve) => {
    const id = nextPermissionId++;
    pendingPermissions.set(id, { resolve, suggestions: options.suggestions });
    options.signal.addEventListener("abort", () => {
      if (pendingPermissions.delete(id)) {
        send({ type: "permission_cancel", id });
        resolve({ behavior: "deny", message: "Permission request was cancelled." });
      }
    });
    send({
      type: "permission_request",
      id,
      tool_use_id: options.toolUseID,
      tool_name: toolName,
      input,
      title: options.title,
      description: options.description,
      has_suggestions: (options.suggestions?.length ?? 0) > 0,
      default_to_no: options.defaultToNo ?? false,
      suppress_always: options.suppressAlwaysAllowRule ?? false,
      agent_id: options.agentID,
    });
  });

/** Accepting releases every held message at once; null then restores whatever applied before. */
async function deliverHeld(): Promise<void> {
  if (!session) return;
  await session.applyFlagSettings({ crossSessionInbound: "accept" });
  await session.applyFlagSettings({ crossSessionInbound: initRequest?.inbound ?? null });
}

let sessionId: string | undefined;
let initRequest: InitRequest | undefined;

async function run(init: InitRequest): Promise<void> {
  initRequest = init;
  sessionId = init.resume ?? init.session_id ?? randomUUID();
  session = query({
    prompt: inbox,
    options: {
      cwd: init.cwd,
      pathToClaudeCodeExecutable: init.claude_path,
      model: init.model,
      permissionMode: init.permission_mode,
      // The SDK refuses bypassPermissions without this; only set it when that mode was chosen.
      allowDangerouslySkipPermissions: init.permission_mode === "bypassPermissions" || undefined,
      resume: init.resume,
      sessionId: init.resume ? undefined : sessionId,
      title: init.resume ? undefined : init.title,
      // Without this the SDK runs with an empty system prompt, not Claude Code's.
      systemPrompt: { type: "preset", preset: "claude_code" },
      includePartialMessages: true,
      promptSuggestions: init.prompt_suggestions ?? true,
      canUseTool,
      settings: init.inbound ? { crossSessionInbound: init.inbound } : undefined,
      extraArgs: {
        // Echo user messages back: how a message from another session (which starts or joins
        // a turn without a prompt from us) reaches the transcript.
        "replay-user-messages": null,
        ...(init.name ? { name: init.name } : {}),
      },
      env: {
        ...process.env,
        // system/session_state_changed: running and idle, whatever started the turn.
        CLAUDE_CODE_EMIT_SESSION_STATE_EVENTS: "1",
      },
    },
  });
  send({ type: "ready", session_id: sessionId });
  session
    .supportedCommands()
    .then((commands) => send({ type: "commands", commands }))
    .catch(() => {});
  for await (const message of session) {
    send({ type: "sdk", message });
  }
}

function handle(request: Inbound): void {
  switch (request.type) {
    case "init":
      if (session) {
        send({ type: "error", message: "Session already initialized." });
        return;
      }
      run(request)
        .catch((err: unknown) => {
          // After an interrupted turn the SDK reports the CLI's normal exit as an error.
          if (closing) return;
          send({ type: "error", message: err instanceof Error ? err.message : String(err) });
        })
        .finally(() => {
          send({ type: "exit" });
          process.exit(0);
        });
      return;
    case "prompt":
      inbox.push(request.text, request.should_query ?? true, request.images);
      return;
    case "set_permission_mode":
      session?.setPermissionMode(request.mode).catch((err: unknown) => {
        send({ type: "error", message: `Couldn't switch to ${request.mode}: ${String(err)}` });
      });
      return;
    case "interrupt":
      session?.interrupt().catch((err: unknown) => {
        send({ type: "error", message: `Interrupt failed: ${String(err)}` });
      });
      return;
    case "rename":
      // Not in the SDK's typings, but on Query since 0.3: the control request the CLI's
      // /rename makes, which also updates the name in the cross-session registry.
      (session as unknown as { renameSession(title: string, sessionId?: string): Promise<void> } | undefined)
        ?.renameSession(request.title, sessionId)
        .catch((err: unknown) => {
          send({ type: "error", message: `Rename failed: ${String(err)}` });
        });
      return;
    case "deliver_held":
      deliverHeld().catch((err: unknown) => {
        send({ type: "error", message: `Couldn't deliver held messages: ${String(err)}` });
      });
      return;
    case "permission_response": {
      const pending = pendingPermissions.get(request.id);
      if (!pending) return;
      pendingPermissions.delete(request.id);
      if (request.behavior === "deny") {
        pending.resolve({ behavior: "deny", message: request.message ?? "The user denied this tool use." });
      } else {
        const updates: PermissionUpdate[] = request.always ? [...(pending.suggestions ?? [])] : [];
        if (request.set_mode) {
          updates.push({ type: "setMode", mode: request.set_mode, destination: "session" });
        }
        pending.resolve({
          behavior: "allow",
          updatedInput: request.updated_input,
          updatedPermissions: updates.length > 0 ? updates : undefined,
        });
      }
      return;
    }
    default:
      send({ type: "error", message: `Unknown request type: ${(request as { type: string }).type}` });
  }
}

export function runSession(): void {
  onLines(
    (request) => handle(request as Inbound),
    // Neovim closing our stdin ends the session.
    () => {
      closing = true;
      inbox.close();
      if (!session) process.exit(0);
    },
    (line) => send({ type: "error", message: `Invalid JSON from Neovim: ${line}` }),
  );
}
